dashboard/cron.py

```diff
@@ -290,58 +290,136 @@ def update_with_bq_access(self):
+                # scientific notation; converting SN to int64 causes
+                # Obi-Wan problems (off by one)
+                user_id_df = pd.read_sql(
+                    'select sis_name as user_login_name,'
+                    'cast(user_id as char) as user_id_str '
+                    f'from user where sis_name in ({login_names})',
+                    engine)
+
+                logger.debug(f'user_id_df:\n'
+                             f'{user_id_df}\n'
+                             f'{user_id_df.dtypes}')
+
+                # combine user login and ID data
+                resource_access_df = pd.merge(
+                    resource_access_df, user_id_df,
+                    on='user_login_name', how='outer')
+
+                # replace real user_id values for missing ones (-1)
+                resource_access_df.loc[
+                    resource_access_df['user_id'] == -1,
+                    'user_id'] = resource_access_df['user_id_str']
+
+                # drops must be in this order; especially dropna() LAST
+                resource_access_df = resource_access_df \
+                    .drop(columns=['user_id_str', 'user_login_name']) \
+                    .dropna()
+
+                logger.debug(f'resource_access_df:\n'
+                             f'{resource_access_df}\n'
+                             f'{resource_access_df.dtypes}')
+            else:
+                resource_access_df = resource_access_df.drop(
+                    columns='user_login_name')
+            resource_access_df = resource_access_df.dropna()
+
             # drop duplicates
-            resource_access_df = resource_access_df.drop_duplicates(["resource_id", "user_id", "access_time"], keep='first')
+            resource_access_df = resource_access_df.drop_duplicates(
+                ['resource_id', 'user_id', 'access_time'], keep='first')
 
-            logger.debug("after drop duplicates, df row number=" + str(resource_access_df.shape[0]))
+            logger.debug('resource_access_df row count (de-duped): '
+                         f'({len(resource_access_df)})')
 
-            logger.debug(resource_access_df)
+            logger.debug(f'resource_access_df:\n'
+                         f'{resource_access_df}\n'
+                         f'{resource_access_df.dtypes}')
 
-            # Because we're pulling all the data down into one query we need to manipulate it a little bit
-            # Make a copy of the access dataframe
+            # Make resource data from resource_access data
             resource_df = resource_access_df.copy(deep=True)
-            # Drop out the columns user and access time from resource data frame
-            resource_df = resource_df.drop(["user_id", "access_time"], axis=1)
-            # Drop out the duplicates
-            resource_df = resource_df.drop_duplicates(["resource_id"])
+            resource_df = resource_df.drop(
+                columns=['user_id', 'access_time', 'course_id'])
```

Comment:
This is where I added the drop of `course_id`.  Because it wasn't being dropped, SQLAlchemy (or perhaps upsert) was adding this column back to the table.  I don't think it's pandas' fault.

Reply:
I actually [changed some of this logic here,](https://github.com/tl-its-umich-edu/my-learning-analytics/pull/1018/files) to not include the column in the original DataFrame. This hasn't been merged yet but it going to conflict with the code.

I thought about dropping these columns but seemed to make more sense and future-proof to just keep the ones I want rather than drop the ones I don't.

Reply:
Thanks for the feedback.  Zhen suggested I merge this PR, so I'll do that.  You can resolve conflicts in your branch.

Reply:
Oh, thanks! :) 

Yeah, that branch needs more work so I can re-do the logic there again.